File: app/web.py
"""Flask アプリとルーティング。

操作はすべて通常のフォーム送信で行い、書き込み後は転送して再送信を防ぐ。
JavaScript を切っていても予約と投票ができる。
"""

from __future__ import annotations

import secrets
from datetime import date
from pathlib import Path

from flask import Flask, flash, g, redirect, render_template, request, url_for

from . import api, clock, config, db, meals, slots
from .meals import Meal

COOKIE = "member_id"
COOKIE_MAX_AGE = 60 * 60 * 24 * 30

MESSAGES = {
    "slot_taken": "この枠はちょうど埋まりました。",
    "past": "その枠は始まっています。",
    "out_of_period": "その日は予約できません。",
    "not_selectable_date": "その日は予約できません。",
    "bad_slot": "その枠は選べません。",
}


def format_go(total: float) -> str:
    """0.5 刻みなので、整数のときは小数点を落とす。"""
    return f"{total:g}"


def _secret_key(db_path: Path) -> str:
    path = Path(db_path).parent / "secret_key"
    if not path.exists():
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(secrets.token_hex(32), encoding="utf-8")
    return path.read_text(encoding="utf-8").strip()


def create_app(config_path=None, db_path=None) -> Flask:
    app = Flask(__name__)
    db_path = Path(db_path) if db_path else config.DEFAULT_DB_PATH
    settings = config.load_settings(config_path)
    app.secret_key = _secret_key(db_path)
    app.config["DB_PATH"] = db_path
    app.config["SETTINGS"] = settings

    conn = db.connect(db_path)
    db.init_schema(conn)
    conn.close()

    # 名簿の名前をそのまま読めるようにする。JSON の規格上 UTF-8 でよい。
    app.json.ensure_ascii = False
    app.register_blueprint(api.bp)
    api.register_errors(app)

    @app.before_request
    def _open_db():
        g.conn = db.connect(app.config["DB_PATH"])

    @app.teardown_request
    def _close_db(exc):
        conn = g.pop("conn", None)
        if conn is not None:
            conn.close()

    @app.context_processor
    def _inject():
        return {"member": current_member(), "config": config}

    # --- 補助 -------------------------------------------------------------

    def current_member():
        raw = request.cookies.get(COOKIE)
        if not raw or not raw.isdigit():
            return None
        return db.get_member(g.conn, int(raw))

    def parse_date(raw: str | None) -> date | None:
        if not raw:
            return None
        try:
            return date.fromisoformat(raw)
        except ValueError:
            return None

    def default_bath_date(now) -> date:
        selectable = slots.selectable_dates(now)
        if not selectable:
            return slots.stay_date(now)
        for day in selectable:
            if slots.has_open_slot(day, now):
                return day
        return selectable[0]

    def bath_context(day: date, now, member):
        reserved = db.reservations_for_date(g.conn, day.isoformat())
        sections = []
        for section in slots.sections_for_date(day):
            columns = []
            for room in config.ROOMS:
                rows = []
                for slot in slots.slot_starts(section, room):
                    row = reserved.get((room, slot))
                    if row is None:
                        started = slots.slot_datetime(day, slot) <= now
                        rows.append(
                            {"slot": slot, "state": "past" if started else "free",
                             "name": None, "rid": None}
                        )
                    else:
                        mine = member is not None and row["member_id"] == member["id"]
                        rows.append(
                            {"slot": slot, "state": "mine" if mine else "taken",
                             "name": row["name"], "rid": row["id"]}
                        )
                columns.append(
                    {
                        "room": room,
                        "label": config.ROOM_LABELS[room],
                        "minutes": config.ROOM_SLOT_MINUTES[room],
                        "rows": rows,
                    }
                )
            sections.append(
                {"key": section, "label": config.SECTION_LABELS[section], "columns": columns}
            )
        tabs = [
            {"value": d.isoformat(), "label": slots.format_day(d), "current": d == day}
            for d in slots.selectable_dates(now)
        ]
        return {"day": day, "day_value": day.isoformat(), "sections": sections, "tabs": tabs}

    def meals_context(now, member):
        registered = db.count_active_members(g.conn)
        cards = []
        for meal in meals.visible_meals(now):
            day = meal.day.isoformat()
            closes = meals.vote_deadline(meal)
            people, total = db.rice_summary(g.conn, day, meal.kind)
            cards.append(
                {
                    "day": day,
                    "kind": meal.kind,
                    "title": meals.format_meal(meal),
                    "state": meals.vote_state(meal, now),
                    "closes": f"{closes.month}/{closes.day} {closes:%H:%M}",
                    "wanted": people,
                    "total": format_go(total),
                    "registered": registered,
                    "voted": member is not None
                    and db.member_vote(g.conn, day, meal.kind, member["id"]),
                }
            )
        return {"cards": cards}

    # --- 名前の選択 -------------------------------------------------------

    @app.get("/")
    def index():
        return redirect(url_for("bath"))

    @app.get("/select")
    def select():
        return render_template("select.html", members=db.list_members(g.conn))

    @app.post("/select")
    def select_post():
        member_id = request.form.get("member_id", "")
        if not member_id.isdigit() or db.get_member(g.conn, int(member_id)) is None:
            flash("名前を選んでください。")
            return redirect(url_for("select"))
        response = redirect(url_for("bath"))
        response.set_cookie(COOKIE, member_id, max_age=COOKIE_MAX_AGE, samesite="Lax")
        return response

    # --- 風呂 -------------------------------------------------------------

    @app.get("/bath")
    def bath():
        member = current_member()
        if member is None:
            return redirect(url_for("select"))
        now = clock.now()
        day = parse_date(request.args.get("date")) or default_bath_date(now)
        return render_template("bath.html", **bath_context(day, now, member))

    @app.get("/fragment/bath")
    def bath_fragment():
        member = current_member()
        now = clock.now()
        day = parse_date(request.args.get("date")) or default_bath_date(now)
        return render_template("_bath_table.html", **bath_context(day, now, member))

    @app.post("/bath/reserve")
    def bath_reserve():
        member = current_member()
        if member is None:
            return redirect(url_for("select"))
        now = clock.now()
        day = parse_date(request.form.get("date"))
        section = request.form.get("section", "")
        room = request.form.get("room", "")
        slot = request.form.get("slot", "")
        if day is None or room not in config.ROOMS:
            flash(MESSAGES["bad_slot"])
            return redirect(url_for("bath"))
        reason = slots.check_reservable(day, section, room, slot, now)
        if reason != "ok":
            flash(MESSAGES[reason])
            return redirect(url_for("bath", date=day.isoformat()))
        result = db.reserve(g.conn, day.isoformat(), section, room, slot, member["id"])
        if result == "already_has":
            held = db.member_reservation(g.conn, day.isoformat(), section, member["id"])
            flash(
                f"{config.SECTION_LABELS[section]}の枠は {held['slot']} の"
                f"{config.ROOM_LABELS[held['room']]}を予約済みです。"
                "取り消してから選んでください。"
            )
        elif result == "slot_taken":
            flash(MESSAGES["slot_taken"])
        return redirect(url_for("bath", date=day.isoformat()))

    @app.post("/bath/cancel")
    def bath_cancel():
        member = current_member()
        if member is None:
            return redirect(url_for("select"))
        raw = request.form.get("reservation_id", "")
        day = parse_date(request.form.get("date"))
        if raw.isdigit() and not db.cancel_reservation(g.conn, int(raw), member["id"]):
            flash("他の人の予約は取り消せません。")
        target = day.isoformat() if day else None
        return redirect(url_for("bath", date=target))

    # --- ごはん -----------------------------------------------------------

    @app.get("/meals")
    def meals_page():
        member = current_member()
        if member is None:
            return redirect(url_for("select"))
        return render_template("meals.html", **meals_context(clock.now(), member))

    @app.get("/fragment/meals")
    def meals_fragment():
        member = current_member()
        return render_template("_meal_cards.html", **meals_context(clock.now(), member))

    def _meal_from_form():
        day = parse_date(request.form.get("date"))
        kind = request.form.get("kind", "")
        if day is None or kind not in config.MEALS or not meals.meal_in_period(kind, day):
            return None
        return Meal(day, kind)

    @app.post("/meals/vote")
    def meals_vote():
        member = current_member()
        if member is None:
            return redirect(url_for("select"))
        meal = _meal_from_form()
        size = request.form.get("size", "")
        if size not in config.RICE_SIZES:
            flash(MESSAGES["bad_slot"])
        elif meal is None or meals.vote_state(meal, clock.now()) != "open":
            flash("受付時間外です。")
        else:
            db.vote(g.conn, meal.day.isoformat(), meal.kind, member["id"], size)
        return redirect(url_for("meals_page"))

    @app.post("/meals/unvote")
    def meals_unvote():
        member = current_member()
        if member is None:
            return redirect(url_for("select"))
        meal = _meal_from_form()
        if meal is None or meals.vote_state(meal, clock.now()) != "open":
            flash("受付時間外です。")
        else:
            db.unvote(g.conn, meal.day.isoformat(), meal.kind, member["id"])
        return redirect(url_for("meals_page"))

    # --- 名簿 -------------------------------------------------------------

    @app.get("/members")
    def members():
        return render_template("members.html", members=db.list_members(g.conn))

    app.jinja_env.globals["current_member"] = current_member
    return app

File: src/web.js
/**
 * 画面のルーティング。
 *
 * 操作はすべて通常のフォーム送信で行い、書き込み後は転送して再送信を防ぐ。
 * JavaScript を切っていても予約と投票ができる。
 */

import path from "node:path";

import ejs from "ejs";
import express from "express";

import { clock } from "./clock.js";
import * as config from "./config.js";
import * as db from "./db.js";
import * as jst from "./jst.js";
import * as meals from "./meals.js";
import * as slots from "./slots.js";

const COOKIE = "member_id";
const COOKIE_MAX_AGE = 60 * 60 * 24 * 30 * 1000;
const NOTICE = "notice";
const NOTICE_MAX_AGE = 10 * 1000;

/**
 * 転送をまたいで出す短い知らせ。
 *
 * Flask の flash は署名付きの session に載せていた。ここでは合図だけを cookie に
 * 置き、文面はサーバ側の表から引く。利用者が cookie を書き換えても、表に無い合図
 * は捨てられるので任意の文字列は画面に出せない。
 */
const NOTICES = {
  slot_taken: "この枠はちょうど埋まりました。",
  past: "その枠は始まっています。",
  out_of_period: "その日は予約できません。",
  not_selectable_date: "その日は予約できません。",
  bad_slot: "その枠は選べません。",
  pick_name: "名前を選んでください。",
  not_yours: "他の人の予約は取り消せません。",
  closed: "受付時間外です。",
};

const view = (name) => path.join(config.VIEW_DIR, `${name}.ejs`);

/** 0.5 刻みなので、整数のときは小数点を落とす。 */
export const formatGo = (total) => String(total);

function readCookies(req) {
  const out = {};
  for (const part of (req.headers.cookie ?? "").split(";")) {
    const at = part.indexOf("=");
    if (at < 0) continue;
    const name = part.slice(0, at).trim();
    if (name in out) continue;
    try {
      out[name] = decodeURIComponent(part.slice(at + 1).trim());
    } catch {
      // 壊れた cookie は無かったことにする。
    }
  }
  return out;
}

/** cookie の合図を文面にする。読めない合図は null。 */
function readNotice(raw) {
  if (!raw) return null;
  if (raw in NOTICES) return NOTICES[raw];

  // 時刻そのものが : を含むため、区切りでは分解せず形ごと照合する。
  const found = /^already_has:([a-z]+):([a-z]+):(\d{2}:\d{2})$/.exec(raw);
  if (found === null) return null;
  const [, section, room, slot] = found;
  if (!config.SECTIONS.includes(section) || !config.ROOMS.includes(room)) return null;
  return (
    `${config.SECTION_LABELS[section]}の枠は ${slot} の` +
    `${config.ROOM_LABELS[room]}を予約済みです。取り消してから選んでください。`
  );
}

const setNotice = (res, code, cookiePath) =>
  res.cookie(NOTICE, code, { maxAge: NOTICE_MAX_AGE, sameSite: "lax", path: cookiePath });

export function createWebRouter(database, basePath = "") {
  const router = express.Router({ strict: true });

  // 自分がぶら下がっている位置を付けた行き先。転送も画面の中のリンクもこれを通す。
  const u = (path) => `${basePath}${path}`;
  const cookiePath = basePath === "" ? "/" : basePath;
  const flash = (res, code) => setNotice(res, code, cookiePath);
  router.use(express.urlencoded({ extended: false, limit: "64kb" }));

  // --- 補助 ---------------------------------------------------------------

  function currentMember(req) {
    const raw = readCookies(req)[COOKIE];
    if (!raw || !/^\d+$/.test(raw)) return null;
    return db.getMember(database, Number(raw));
  }

  async function page(req, res, name, locals) {
    const cookies = readCookies(req);
    const notice = readNotice(cookies[NOTICE]);
    if (cookies[NOTICE] !== undefined) res.clearCookie(NOTICE, { path: cookiePath });

    const data = {
      config,
      formatGo,
      base: basePath,
      member: currentMember(req),
      notice,
      tab: "",
      fragment: "",
      ...locals,
    };
    const body = await ejs.renderFile(view(name), data);
    res.type("html").send(await ejs.renderFile(view("base"), { ...data, body }));
  }

  const fragment = (res, name, locals) =>
    ejs
      .renderFile(view(name), { config, formatGo, base: basePath, ...locals })
      .then((html) => res.type("html").send(html));

  function defaultBathDate(now) {
    const selectable = slots.selectableDates(now);
    if (selectable.length === 0) return slots.stayDate(now);
    return selectable.find((day) => slots.hasOpenSlot(day, now)) ?? selectable[0];
  }

  function bathContext(day, now, member) {
    const reserved = db.reservationsForDate(database, day);
    const sections = slots.sectionsForDate(day).map((section) => ({
      key: section,
      label: config.SECTION_LABELS[section],
      columns: config.ROOMS.map((room) => ({
        room,
        label: config.ROOM_LABELS[room],
        minutes: config.ROOM_SLOT_MINUTES[room],
        rows: slots.slotStarts(section, room).map((slot) => {
          const row = reserved.get(db.slotKey(room, slot));
          if (row === undefined) {
            const started = slots.slotInstant(day, slot) <= now;
            return { slot, state: started ? "past" : "free", name: null, rid: null };
          }
          const mine = member !== null && row.member_id === member.id;
          return { slot, state: mine ? "mine" : "taken", name: row.name, rid: row.id };
        }),
      })),
    }));
    const tabs = slots.selectableDates(now).map((value) => ({
      value,
      label: slots.formatDay(value),
      current: value === day,
    }));
    return { day, dayValue: day, sections, tabs };
  }

  function mealsContext(now, member) {
    const registered = db.countActiveMembers(database);
    const cards = meals.visibleMeals(now).map((meal) => {
      const { people, total } = db.riceSummary(database, meal.day, meal.kind);
      return {
        day: meal.day,
        kind: meal.kind,
        title: meals.formatMeal(meal),
        state: meals.voteState(meal, now),
        closes: jst.formatClock(meals.voteDeadline(meal)),
        wanted: people,
        total: formatGo(total),
        registered,
        voted: member === null ? null : db.memberVote(database, meal.day, meal.kind, member.id),
      };
    });
    return { cards };
  }

  const askedDate = (req) => jst.parseDay(req.query.date ?? req.body?.date ?? "");

  // Flask は経路が別の method 用に登録されていれば 405 を返す。画面側もそれに揃える。
  const wrongMethod = (req, res) => res.status(405).type("text").send("method が違う");

  // --- 名前の選択 ---------------------------------------------------------

  router.route("/").get((req, res) => res.redirect(u("/bath"))).all(wrongMethod);

  router
    .route("/select")
    .get((req, res) => page(req, res, "select", { members: db.listMembers(database) }))
    .post((req, res) => {
      const raw = String(req.body.member_id ?? "");
      if (!/^\d+$/.test(raw) || db.getMember(database, Number(raw)) === null) {
        flash(res, "pick_name");
        return res.redirect(u("/select"));
      }
      res.cookie(COOKIE, raw, { maxAge: COOKIE_MAX_AGE, sameSite: "lax", path: cookiePath });
      return res.redirect(u("/bath"));
    })
    .all(wrongMethod);

  // --- 風呂 ---------------------------------------------------------------

  router.route("/bath").get((req, res) => {
    const member = currentMember(req);
    if (member === null) return res.redirect(u("/select"));
    const now = clock.now();
    const day = askedDate(req) ?? defaultBathDate(now);
    return page(req, res, "bath", {
      ...bathContext(day, now, member),
      tab: "bath",
      fragment: u(`/fragment/bath?date=${day}`),
    });
  }).all(wrongMethod);

  router.route("/fragment/bath").get((req, res) => {
    const now = clock.now();
    const day = askedDate(req) ?? defaultBathDate(now);
    return fragment(res, "_bath_table", bathContext(day, now, currentMember(req)));
  }).all(wrongMethod);

  router.route("/bath/reserve").post((req, res) => {
    const member = currentMember(req);
    if (member === null) return res.redirect(u("/select"));
    const now = clock.now();
    const day = askedDate(req);
    const section = String(req.body.section ?? "");
    const room = String(req.body.room ?? "");
    const slot = String(req.body.slot ?? "");

    if (day === null || !config.ROOMS.includes(room)) {
      flash(res, "bad_slot");
      return res.redirect(u("/bath"));
    }
    const reason = slots.checkReservable(day, section, room, slot, now);
    if (reason !== "ok") {
      flash(res, reason);
      return res.redirect(u(`/bath?date=${day}`));
    }
    const result = db.reserve(database, day, section, room, slot, member.id);
    if (result === "already_has") {
      const held = db.memberReservation(database, day, section, member.id);
      flash(res, `already_has:${section}:${held.room}:${held.slot}`);
    } else if (result === "slot_taken") {
      flash(res, "slot_taken");
    }
    return res.redirect(u(`/bath?date=${day}`));
  }).all(wrongMethod);

  router.route("/bath/cancel").post((req, res) => {
    const member = currentMember(req);
    if (member === null) return res.redirect(u("/select"));
    const raw = String(req.body.reservation_id ?? "");
    const day = askedDate(req);
    if (/^\d+$/.test(raw) && !db.cancelReservation(database, Number(raw), member.id)) {
      flash(res, "not_yours");
    }
    return res.redirect(u(day === null ? "/bath" : `/bath?date=${day}`));
  }).all(wrongMethod);

  // --- ごはん -------------------------------------------------------------

  router.route("/meals").get((req, res) => {
    const member = currentMember(req);
    if (member === null) return res.redirect(u("/select"));
    return page(req, res, "meals", {
      ...mealsContext(clock.now(), member),
      tab: "meals",
      fragment: u("/fragment/meals"),
    });
  }).all(wrongMethod);

  router
    .route("/fragment/meals")
    .get((req, res) => fragment(res, "_meal_cards", mealsContext(clock.now(), currentMember(req))))
    .all(wrongMethod);

  function mealFromForm(req) {
    const day = jst.parseDay(req.body.date ?? "");
    const kind = String(req.body.kind ?? "");
    if (day === null || !config.MEALS.includes(kind) || !meals.mealInPeriod(kind, day)) return null;
    return { day, kind };
  }

  router.route("/meals/vote").post((req, res) => {
    const member = currentMember(req);
    if (member === null) return res.redirect(u("/select"));
    const meal = mealFromForm(req);
    const size = String(req.body.size ?? "");
    if (!config.RICE_SIZES.includes(size)) {
      flash(res, "bad_slot");
    } else if (meal === null || meals.voteState(meal, clock.now()) !== "open") {
      flash(res, "closed");
    } else {
      db.vote(database, meal.day, meal.kind, member.id, size);
    }
    return res.redirect(u("/meals"));
  }).all(wrongMethod);

  router.route("/meals/unvote").post((req, res) => {
    const member = currentMember(req);
    if (member === null) return res.redirect(u("/select"));
    const meal = mealFromForm(req);
    if (meal === null || meals.voteState(meal, clock.now()) !== "open") {
      flash(res, "closed");
    } else {
      db.unvote(database, meal.day, meal.kind, member.id);
    }
    return res.redirect(u("/meals"));
  }).all(wrongMethod);

  // --- 名簿 ---------------------------------------------------------------

  router
    .route("/members")
    .get((req, res) => page(req, res, "members", { members: db.listMembers(database) }))
    .all(wrongMethod);

  return router;
}
